### PR TITLE
Added optional properties for Window aggregate definition

### PR DESCRIPTION
**Pull request by udurbha**
Added additional properties which can be set if frame type is row or range for Windows Aggregation.The values are set from window aggregation repo based on the window frame type value which will be taken from UI.

### CODE REVIEW

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/aggregation/WindowAggregationDefinition.java [hunk 1]
```diff
@@ -33,15 +33,30 @@ public class WindowAggregationDefinition {
   private Map<String, Expression> selectExpressions;
   private List<OrderByExpression> orderByExpressions;
   private WindowFrameType windowFrameType;
+  private String numberOfPartitions;
```
Comment (fernst) [inline cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/aggregation/WindowAggregationDefinition.java]: Number of partitions is not required to generate the SQL statement. Please remove.

**Review by fernst — CHANGES_REQUESTED** [review]

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/aggregation/WindowAggregationDefinition.java (hunk 1 above)
Comment (udurbha) [reply cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/aggregation/WindowAggregationDefinition.java]: Okay

**Review by fernst — APPROVED** [review]

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/aggregation/WindowAggregationDefinition.java [hunk 2]
```diff
@@ -33,15 +34,27 @@ public class WindowAggregationDefinition {
...
+  private boolean unboundedPreceding;
+  private boolean unboundedFollowing;
+  private String preceding;
+  private String following;
 
   private WindowAggregationDefinition(Map<String, Expression> selectExpressions,
                                       List<Expression> partitionExpressions,
-                                      List<OrderByExpression> orderByExpressions,
-                                      WindowFrameType windowFrameType) {
+                                      @Nullable List<OrderByExpression> orderByExpressions,
```
Comment (fernst) [inline cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/aggregation/WindowAggregationDefinition.java:44]: Also add `@Nullable` to this getter

Comment (fernst) [reply cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/aggregation/WindowAggregationDefinition.java:44]: Nevermind, saw you added it

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/aggregation/WindowAggregationDefinition.java [hunk 3]
```diff
@@ -211,7 +306,7 @@ public Builder windowFrameType(WindowFrameType windowFrameType) {
      * Any existing order by expression list is overwritten.
      * @return a {@link Builder} with the currently built {@link WindowAggregationDefinition}.
      */
-    public Builder orderBy(Expression expression, OrderBy orderBy) {
+    public Builder orderBy(@Nullable Expression expression,  @Nullable OrderBy orderBy) {
```
Comment (fernst) [inline cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/aggregation/WindowAggregationDefinition.java]: These shouldn't be nullable. The order by expression and order parameter are required when defining the ordering criteria.

Also, if the list is null, maybe you should initialize the list before adding an expression

**Review by fernst — CHANGES_REQUESTED** [review]

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/aggregation/WindowAggregationDefinition.java (hunk 3 above)
Comment (udurbha) [reply cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/aggregation/WindowAggregationDefinition.java]: In the existing windowsaggregation plugin, order by field is accepting even if one of the fields(column name or order by asc or desc) is entered,Based on that added @nullable.Will raise jira to fix the existing behaviour.Removed nullable from this method

**Review by fernst — APPROVED** [review]

**Pull request merged**